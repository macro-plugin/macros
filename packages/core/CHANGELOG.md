# @macro-plugin/core

## 1.2.0

### Minor Changes

- [#33](https://github.com/macro-plugin/macros/pull/33) [`8c94045`](https://github.com/macro-plugin/macros/commit/8c9404545d207b708ed433328bd383488a3ed0d4) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: new cursor context method

- [#33](https://github.com/macro-plugin/macros/pull/33) [`b45063f`](https://github.com/macro-plugin/macros/commit/b45063ff784e513380f43bf8425c4f72eaa5c0b4) Thanks [@voorjaar](https://github.com/voorjaar)! - refactor: rename span to dummySpan

## 1.1.4

### Patch Changes

- [#29](https://github.com/macro-plugin/macros/pull/29) [`e3ce376`](https://github.com/macro-plugin/macros/commit/e3ce37696e0a21bdddb56804f7dbcfa904f5fd35) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: support resolveTs and swcTransform options

## 1.1.3

## 1.1.2

### Patch Changes

- [#23](https://github.com/macro-plugin/macros/pull/23) [`294f675`](https://github.com/macro-plugin/macros/commit/294f67574c5ff168bdeb0a499ea4ab2fd57579cf) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: support include, exclude, exclude, extensions, tsconfig options

## 1.1.1

### Patch Changes

- [#21](https://github.com/macro-plugin/macros/pull/21) [`b4be79b`](https://github.com/macro-plugin/macros/commit/b4be79be09e8790b95c6a2e9b07dde4ba3822f39) Thanks [@voorjaar](https://github.com/voorjaar)! - chore: update to swc 1.3.55

- [#21](https://github.com/macro-plugin/macros/pull/21) [`df8dcc0`](https://github.com/macro-plugin/macros/commit/df8dcc0ab7f3e09f1157c5c2a7cda4a7f5367991) Thanks [@voorjaar](https://github.com/voorjaar)! - fix: use swc Plugin type exports

## 1.1.0

### Minor Changes

- [#16](https://github.com/macro-plugin/macros/pull/16) [`4155bb4`](https://github.com/macro-plugin/macros/commit/4155bb4de7968a83e62203411bceae6b0602637f) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: support loading external macros

### Patch Changes

- [#16](https://github.com/macro-plugin/macros/pull/16) [`6b25910`](https://github.com/macro-plugin/macros/commit/6b25910567e910b7c71c79646f8569a2f3927be6) Thanks [@voorjaar](https://github.com/voorjaar)! - fix: types for $Quote template macro

- [#16](https://github.com/macro-plugin/macros/pull/16) [`8dd4939`](https://github.com/macro-plugin/macros/commit/8dd493997931d8d91a82ffb1785927d425c17c61) Thanks [@voorjaar](https://github.com/voorjaar)! - fix: tracker failed to load expression macro when using import

- [#16](https://github.com/macro-plugin/macros/pull/16) [`14efcb2`](https://github.com/macro-plugin/macros/commit/14efcb2c6461ab3f5d78e0599ec74b422085ce1d) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: support access config from macro context

- [#16](https://github.com/macro-plugin/macros/pull/16) [`7787022`](https://github.com/macro-plugin/macros/commit/7787022f657b7a79cb18a1e8ba947ae2eaeb682e) Thanks [@voorjaar](https://github.com/voorjaar)! - fix: mark proxy as macro_plugin

- [#16](https://github.com/macro-plugin/macros/pull/16) [`a17b055`](https://github.com/macro-plugin/macros/commit/a17b055d356c285b394add56192c80077ebde2c2) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: new AST types exported from swc

## 1.0.3

### Patch Changes

- [#12](https://github.com/macro-plugin/macros/pull/12) [`07961a0`](https://github.com/macro-plugin/macros/commit/07961a03b6e82080a2b8c8ab2626c187c34f912e) Thanks [@voorjaar](https://github.com/voorjaar)! - refactor: move createLabeledExpr and createLabeledBlock to labeled

- [#12](https://github.com/macro-plugin/macros/pull/12) [`f1e40ea`](https://github.com/macro-plugin/macros/commit/f1e40ead32636d4f2d43c442e70cc208e2d43b28) Thanks [@voorjaar](https://github.com/voorjaar)! - fix: decorator in arrowFunction with variableDeclarator should keep name

- [#12](https://github.com/macro-plugin/macros/pull/12) [`556ca2d`](https://github.com/macro-plugin/macros/commit/556ca2d9addaf36ac84da8c8ea7b5bc465e174b7) Thanks [@voorjaar](https://github.com/voorjaar)! - feat: support labeled function decorator

## 1.0.2

### Patch Changes

- [#9](https://github.com/macro-plugin/macros/pull/9) [`2c64787`](https://github.com/macro-plugin/macros/commit/2c647875182c5fc6ca41c9e72587a08307ba90ec) Thanks [@voorjaar](https://github.com/voorjaar)! - chore: upgrade to swc 1.3.53

## 1.0.1

### Patch Changes

- 98425cf: build: no longer publish iife
