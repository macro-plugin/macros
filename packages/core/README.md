# @macro-plugin/core

Macro system for JavaScript and TypeScript.

## Install

```sh
# if you use npm
npm i -D @macro-plugin/core
# if you use pnpm
pnpm i -D @macro-plugin/core
# if you use yarn
yarn add -D @macro-plugin/core
```

## Why

Have you ever used Rust? Rust has an impressive macro system. I'm wondering why the JavaScript world can't have one. Now you got it.

## How it works

It uses [swc](https://swc.rs/)'s parser and printer, and does not make any conversions other than macors to your project by default. Of course, it can also be used as a swc plugin, so we can transform typescript and jsx.

## Features

- A new walker for macros, inspired by [estree-walker](https://github.com/Rich-Harris/estree-walker)
- Built-in variable tracker plugin, which makes it easier to convert variables.
- Very simple yet powerful APIs for creating various types of macros.
- Factory runtime and macros for simplify creating swc AST.
- Support automatic generation of macro declarations file. (macros.d.ts)
- Many common built-in macros.
- Support all web framework intergrations.

## Built-in macros

The project is still in the beta stage, and there are more macros under development.

### Expression Macros

Expression macros may be a very familiar macro type. It converts a `CallExpression` into any other Expression.

#### `$Eval`

compile time eval.

#### `$Ast`

convert anything into an ast.

#### `$Env`

read env and replace with result.

#### `$Stringify`

convert any expression or type into string

#### `$Span`

get a span like `[line, column]`

#### `$Line`

get current line number

#### `$Column`

get current column number

#### `$ID`

generate unique id based on `[line, column, filename]`

#### `$Include`

include another `js, ts` file into current script.

#### `$IncludeStr`

include a string from any path

#### `$IncludeJSON`

include a json, also support include a json value with key.

#### `$WriteFile`

compile time write to file

#### `$Concat`

compile time concat string

#### `$Expr`

turn code into an expression

#### `$Quote`

a template macro for create macro

#### `$UnImplemented`

mark code as unImplemented, log error when compile, throw error in client side.

#### `$Todo`

mark code as todo, log warning when compile, throw error in client side.

#### `$UnReachable`

mark code as unreachable, throw error in client side.

...

### Labeled Macros

LabeledMacros are very interesting feature. It even allows you to invent your own grammar.
For example:

```js
debug: {
  console.log("something")
}
```

or creating states:

```js
state: {
  var count = 0
  var name = "macro"
}
```

or using function decorator:

```js
function myFunc() {
  decorator: [yourFunctionDecorator]

  doSomething()
}
```

you can even create macro with it.

```ts
macro: {
  var __DEV__ = true
  
  var add = $ExprMacro<(a: number, b: number) => number>(function(args) {
    const a = this.printExpr(args[0])
    const b = this.printExpr(args[1])

    if (+a < 0) return args[1]
    return this.parseExpr(`${a} + ${b}`)
  })
}

if (__DEV__) {
  const a = add(1, 2)
}
```

transform to

```js
if (true) {
  const a = 1 + 2
}
```


To use Labeled Macros, you may need some extra config for typescript and eslint.

#### Typescript

```json
{
  "compilerOptions": {
    // ...
    "allowUnusedLabels": true,
  },
}
```

#### Eslint

```js
module.exports = {
  rules: {
    // ...
    "no-labels": 0,
  }
}
```

## API

### `transform/transformAst/transformAsync`

transform code with macro plugins.

### `createSwcPlugin`

use as swc plugin.

### `walk`

walk an ast node or node list

### `print/printAsync/printExpr/printType`

print ast to code

### `parse/parseAsync/parseExpr/parseType`

parse code to ast

### `defineConfig`

define config with types

### `createMacro`

create a macro

### `createLitMacro`

create a literal macro

### `createExprMacro`

create an expression macro

### `createTypeMacro`

create a type macro

### `createTmplMacro`

create a template macro

### `createLabeledMacro`

create a labeled macro

### `isMacroPlugin`

check if input is a macro plugin

## FAQ

### LabeldMacros confusing with JavaScript

Some people may argue about labeldMacros, confusing the original semantics of JavaScript. But in fact, `labeledStatement` and `var` are both forgotten or discarded syntax of JavaScript. So we bring something new to it. Personally, I don't think this is a bad thing.

### Compare with babel

In fact, the beginning of this project based on babel's parser and generator. When I switched to swc, I did a simple test and found that the speed was at least tripled, so it must be faster.

This project is based on the AST type of swc and is not compatible with Babel's AST. And the project focuses on transforming macros. If you need the support of something like browserlist, you need to use it with swc.

### Compare with swc

You can regard it as an enhanced version of swc with macro support, as well as better aapi. However, we do not use the `.swcrc` configuration file. The json configuration is not friendly for macros, and you should define all your swc configuration in `macros.config.js` or `macros.config.ts`.

## License

[MIT](https://github.com/macro-plugin/macros/blob/main/LICENSE)

Copyright (c) 2023, Raven Satir