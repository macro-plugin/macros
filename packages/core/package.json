{
  "name": "@macro-plugin/core",
  "version": "1.0.1",
  "description": "Macro system for JavaScript and TypeScript.",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "./dist/index.d.ts",
  "bin": {
    "macros": "dist/cli.js"
  },
  "files": [
    "dist",
    "CHANGELOG.md"
  ],
  "keywords": [
    "macro",
    "macros",
    "cli",
    "macro-plugin",
    "macro-system"
  ],
  "scripts": {
    "build": "rollup -c ../../rollup.config.mjs"
  },
  "dependencies": {
    "@swc/core": "^1.3.40"
  },
  "homepage": "https://github.com/macro-plugin/macros/tree/main/packages/core",
  "repository": {
    "type": "git",
    "url": "https://github.com/macro-plugin/macros.git"
  },
  "sideEffects": false,
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./cli": {
      "require": "./dist/cli.js"
    },
    "./register": {
      "require": "./dist/register.js"
    }
  },
  "author": "Raven Satir",
  "license": "MIT"
}
