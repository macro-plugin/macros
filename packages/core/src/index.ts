export { createLabeledBlock } from "./block"
export { createLabeledExpr } from "./expr"
export { createSwcPlugin, transform, transformAst, transformAsync } from "./transform"
export { print, printAsync, printExpr, printExprAsync } from "./print"
export { parse, parseAsync, parseExpr, parseType } from "./parse"
export { walk, Walker } from "./walk"
export { macro, $Macro, $LitMacro, $ExprMacro, $TmplMacro, $TypeMacro, $LabeledMacro } from "./macro"
export * from "./api"
export * from "./lib"
export * from "./types"
export * from "./utils"
