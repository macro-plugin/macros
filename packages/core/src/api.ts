import { ArrowFunctionExpression, Expression, FunctionDeclaration, FunctionExpression, Invalid, Param, TsType } from "@swc/core"
import type { BaseNode, ExprMacro, LabeledMacro, MacroPlugin, WalkContext } from "./types"

import { isRegExp } from "./utils"
import { parseExpr } from "./parse"
import { walk } from "./walk"

export function createMacro (macro: MacroPlugin) {
  return macro
}

function createLit (this: WalkContext, value: unknown): BaseNode {
  if (value === undefined) {
    return {
      type: "Identifier",
      span: {
        start: 0,
        end: 0,
        ctxt: 2
      },
      value: "undefined",
      optional: false
    }
  }
  if (value && typeof value === "object") {
    if (isRegExp(value)) return this.parseExpr(value.toString())
    if ("span" in value && "type" in value) return value as BaseNode
  }
  if (typeof value === "function") return this.parseExpr(value.toString())
  return this.parseExpr(JSON.stringify(value))
}

export function createLitMacro(map: Record<string, unknown>): MacroPlugin;
export function createLitMacro<T>(key: string, value: T): MacroPlugin;
export function createLitMacro (arg: string | Record<string, unknown>, value?: unknown): MacroPlugin {
  return createMacro(typeof arg === "string"
    ? {
      Identifier (ast) {
        if (ast.value === arg && !this.track(arg)) return createLit.apply(this, [value])
      }
    }
    : {
      Identifier (ast) {
        if (ast.value in arg && !this.track(ast.value)) return createLit.apply(this, [arg[ast.value]])
      }
    })
}

function flatExpr (f: Function, args: Expression[], typeParams?: TsType[], optional = false): Expression {
  const ast = parseExpr(f.toString()) as FunctionDeclaration | FunctionExpression | ArrowFunctionExpression
  const params: Record<string, { value?: Expression }> = {}

  ast.params.forEach((p, i) => {
    const pat = ast.type === "ArrowFunctionExpression" ? p : ((p as Param).pat)
    if (pat.type === "Identifier") {
      params[pat.value] = args[i] ? { value: args[i] } : {}
    } else if (pat.type === "AssignmentPattern") {
      if (pat.left.type === "Identifier") {
        params[pat.left.value] = { value: args[i] || pat.right }
      }
    }
  })

  let output: Invalid | Expression = {
    type: "Invalid",
    span: {
      start: 0,
      end: 0,
      ctxt: 0
    }
  }

  if (ast.body) {
    walk(ast.body, {
      // @ts-ignore
      enter (ast: BaseNode) {
        if (ast.type === "Identifier" && ast.value in params) {
          const v = params[ast.value].value
          if (v) this.replace(v)
        }
      },
      // @ts-ignore
      leave (ast: BaseNode) {
        if (ast.type === "ReturnStatement" && ast.argument) {
          output = ast.argument
        }
      }
    })

    if (ast.body.type !== "BlockStatement") output = ast.body
  }

  return output
}

export function createExprMacro (name: string, f: Function | ExprMacro | { enter?: ExprMacro, leave?: ExprMacro }): MacroPlugin {
  if (typeof f === "object") {
    return createMacro({
      CallExpression: {
        enter (ast) {
          if (f.enter && ast.callee.type === "Identifier" && ast.callee.value === name && !this.track(ast.callee.value)) {
            return f.enter.apply(this, [ast.arguments.map(i => i.expression), ast.typeArguments?.params, ast.callee.optional])
          }
        },
        leave (ast) {
          if (f.leave && ast.callee.type === "Identifier" && ast.callee.value === name && !this.track(ast.callee.value)) {
            return f.leave.apply(this, [ast.arguments.map(i => i.expression), ast.typeArguments?.params, ast.callee.optional])
          }
        }
      }
    })
  }
  return createMacro({
    CallExpression (ast) {
      if (ast.callee.type === "Identifier" && ast.callee.value === name && !this.track(ast.callee.value)) {
        const args = ast.arguments.map(i => i.expression)
        const tys = ast.typeArguments?.params
        return f.toString().startsWith("function") ? (f as ExprMacro).apply(this, [args, tys, ast.callee.optional]) : flatExpr(f, args, tys, ast.callee.optional)
      }
    }
  })
}

export function createTypeMacro () {

}

export function createTmplMacro (tag: string,) {

}

export function createLabeledMacro (label: string, f: LabeledMacro | {
  enter?: LabeledMacro,
  leave?: LabeledMacro,
}) {
  return createMacro({
    LabeledStatement: {
      enter (ast, parent, prop, index) {
        if (ast.label.value === label) return (typeof f === "object" ? f.enter : f)?.apply(this, [ast.body, parent, prop, index])
      },
      leave (ast, parent, prop, index) {
        if (ast.label.value === label && typeof f === "object") return f.leave?.apply(this, [ast.body, parent, prop, index])
      }
    }
  })
}
