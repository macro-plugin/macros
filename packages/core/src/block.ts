import { ArrowFunctionExpression, ExpressionStatement } from "@swc/core";

import { MacroPlugin } from "./types";
import { createMacro } from "./api";

/**
 * Create a macro plugin that converts `labeled: {}` or `labeled: (...args) => {}` to `$specifier((...args) => {})`,
 * and also `import { $specifier } from $source`
 * @param specifier - the function name
 * @param source - the module path
 * @param allowParams - allow convert the input array to params, default is false.
 * @returns - A labeled macro plugin
 */
export const createLabeledBlock: ((label: string, specifier: string, source: string, allowParams?: boolean) => MacroPlugin) = (label, specifier, source, allowParams = false) => {
  return createMacro({
    LabeledStatement(ast) {
      if (ast.label.value != label) return;
      this.import([{ name: specifier }], source);
      if (ast.body.type == "BlockStatement") {
        return ({
          type: "ExpressionStatement",
          expression: {
            type: 'CallExpression',
            callee: {
              type: 'Identifier',
              value: specifier,
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            arguments: [
              {
                "expression": {
                  type: 'ArrowFunctionExpression',
                  generator: false,
                  async: false,
                  params: [],
                  body: ast.body,
                  span: {
                    start: 0,
                    end: 0,
                    ctxt: 0
                  }
                } as ArrowFunctionExpression
              }
            ],
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          },
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          }
        }) as ExpressionStatement
      } else if (ast.body.type == "ExpressionStatement" && ast.body.expression.type == 'ArrowFunctionExpression') {
        return ({
          type: "ExpressionStatement",
          expression: {
            type: 'CallExpression',
            callee: {
              type: 'Identifier',
              value: specifier,
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            arguments: [
              {
                expression: ast.body.expression
              }
            ],
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          },
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          }
        }) as ExpressionStatement
      } else if (allowParams && ast.body.type == "ExpressionStatement" && ast.body.expression.type == 'ArrayExpression') {
        return ({
          type: "ExpressionStatement",
          expression: {
            type: 'CallExpression',
            callee: {
              type: 'Identifier',
              value: specifier,
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            arguments: ast.body.expression.elements,
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          },
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          }
        }) as ExpressionStatement
      } else {
        throw new Error('this macro only accept a ArrowFunction or BlockStatement' + (allowParams ? ' or an ArrayExpression' : ''))
      }
    }
  })
}
