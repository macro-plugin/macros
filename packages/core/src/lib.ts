import { Expression, Identifier } from "@swc/core"
import { createExprMacro, createTmplMacro } from "./api"
import { evalAst, evalExpr, span } from "./utils"

import { printExpr } from "./print"
import { walk } from "./walk"

export var $Eval = createExprMacro("$Eval", function (args) {
  const t = args[0].type

  if (t === "StringLiteral") return this.parseExpr(evalExpr(args[0].value))
  if (t === "TemplateLiteral") {
    const track = (name: string) => this.track(name)
    walk(args[0], {
      enter (node) {
        if (node.type === "Identifier") {
          const v = track((node as Identifier).value)
          if (v && v.value) this.replace(v.value)
        }
      }
    })

    return this.parseExpr(evalExpr(evalExpr(this.printExpr(args[0]))))
  }

  if (t === "FunctionExpression" || t === "ArrowFunctionExpression") return this.parseExpr(evalExpr(`(${this.printExpr(args[0])})(${args.slice(1).map(i => this.printExpr(i)).join(",")})`))
  return this.parseExpr(evalAst(args[0]).toString())
}, "(<T>(expr: string) => T) & (<T>(expr: T) => T) & (<F extends (...args: any) => any>(expr: F, ...args: Parameters<F>) => ReturnType<F>)")

export const printAst = (ast: object) => JSON.stringify(ast).replace(/("(start|end|ctxt)":\s*)(\d+)/g, "$10")

export var $Ast = createExprMacro("$Ast", function (args) {
  return this.parseExpr(printAst(args[0].type === "StringLiteral" ? this.parseExpr(args[0].value) : args[0]))
}, '<T>(expr: T) => import("@swc/core").Expression')

export var $Env = createExprMacro("$Env", function (args, typeArgs) {
  if (args[0].type !== "StringLiteral") throw new Error("$Env macro only support StringLiteral as input.")
  const value = process.env[args[0].value] ?? ""

  if (typeArgs && typeArgs?.[0].type === "TsKeywordType") {
    switch (typeArgs[0].kind) {
      case "boolean":
        return {
          type: "BooleanLiteral",
          value: !!value,
          span
        }
      case "number":
        return {
          type: "NumericLiteral",
          value: +value,
          span
        }
    }
  }

  return {
    type: "StringLiteral",
    span,
    value
  }
}, "<R>(key: string) => R")

export var $Stringify = createExprMacro("$Stringify", function (args, typeArgs) {
  return {
    type: "StringLiteral",
    span,
    value: typeArgs && typeArgs.length > 0 ? this.printType(typeArgs[0]) : this.printExpr(args[0])
  }
}, "((expr: any) => string) & (<T>() => string)")

export const printTmpl = (strings: string[], exprs: Expression[]) => strings.reduce((query, queryPart, i) => {
  const valueExists = i < exprs.length
  const text = query + queryPart

  return valueExists ? text + printExpr(exprs[i]).code : text
}, "")

export const printRawTmpl = (strings: string[], exprs: unknown[]) => strings.reduce((query, queryPart, i) => {
  const valueExists = i < exprs.length
  const text = query + queryPart

  return valueExists ? text + exprs[i] : text
}, "")

function walkObject (obj: object | null, onEnter: (k: string, v: unknown, parent: Record<string, any>) => unknown, grand?: Record<string, any>, grandKey?: string) {
  if (obj && typeof obj === "object") {
    const allKeys = Object.keys(obj)
    for (let i = 0; i < allKeys.length; i++) {
      const k = allKeys[i]
      const v = obj[k as keyof typeof obj]

      const r = onEnter(k, v, obj)
      if (r != null && grand) {
        grand[grandKey!] = r
        break
      }

      if (typeof v === "object") {
        walkObject(v, onEnter, obj, k)
      }
    }
  }
  return obj
}

export var $Expr = createTmplMacro("$Expr", function (strings, ...exprs) {
  const exprMarkers: string[] = exprs.map((_, i) => "_macro_marker_" + i + "_")
  const options = { syntax: "typescript", tsx: true } as const
  const ast = this.parseExpr(printRawTmpl(strings, exprMarkers), options)

  walkObject(ast, (k, v, parent) => {
    if (k === "span") {
      parent[k] = span
      return
    }

    if (typeof v === "string") {
      const i = exprMarkers.findIndex((m) => m === v)
      if (i !== -1) return "__macro$$Start__" + this.printExpr(exprs[i]) + "__macro$$End__"
    }
  })

  const expr = JSON.stringify(ast, undefined, 2).replace(/("__macro\$\$Start__)|(__macro\$\$End__")/g, "").replace(/\\"/g, '"')

  return this.parseExpr(expr)
}, "import(\"@swc/core\").Expression")

export var $Quote = createTmplMacro("$Quote", function (strings, ...exprs) {
  const exprMarkers: string[] = exprs.map((_, i) => "_macro_marker_" + i + "_")
  const options = { syntax: "typescript", tsx: true } as const
  const ast = this.parse(printRawTmpl(strings, exprMarkers), options)

  walkObject(ast, (k, v, parent) => {
    if (k === "span") {
      parent[k] = span
      return
    }

    if (typeof v === "string") {
      const i = exprMarkers.findIndex((m) => m === v)
      if (i !== -1) return "__macro$$Start__" + this.printExpr(exprs[i]) + "__macro$$End__"
    }
  })

  const expr = JSON.stringify(ast, undefined, 2).replace(/("__macro\$\$Start__)|(__macro\$\$End__")/g, "").replace(/\\"/g, '"')

  return this.parseExpr(expr)
}, "(import(\"@swc/core\").Expression)[]")
