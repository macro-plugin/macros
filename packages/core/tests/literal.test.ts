import { createLitMacro, createSwcPlugin, parseExpr, transform } from "../src"
import { parseSync, transformSync as swcTransform } from "@swc/core"

test("create literal macro", () => {
  const code = `
  if (__DEV__) {
    console.log('development')
  }
  `

  expect(transform(code, { macros: [createLitMacro("__DEV__", true)] }).code).toMatchSnapshot()
})

test("use as swc plugin", () => {
  const plugin = createSwcPlugin({ macros: [createLitMacro("__DEV__", true)] })

  const code = `
  if (__DEV__) {
    console.log('development')
  }
  `

  expect(swcTransform(code, { plugin, jsc: { parser: { syntax: "typescript" } } }).code).toMatchSnapshot()
})

test("use with swc transform ast", () => {
  const plugin = createSwcPlugin({ macros: [createLitMacro("__DEV__", true)] })

  const code = `
  let a: string = "Hello"

  if (__DEV__) {
    console.log('development')
  }
  `

  const program = plugin(parseSync(code, { syntax: "typescript" }))

  expect(swcTransform(program, { jsc: { parser: { syntax: "typescript" } } }).code).toMatchSnapshot()
})

test("should pass overwrited variable", () => {
  const code = `
  const before = __DEV__

  function testDev(__DEV__) {
    if (__DEV__) {
      console.log('development')
    }
  }

  const after = __DEV__
  `
  expect(transform(code, { macros: [createLitMacro("__DEV__", true)] }).code).toMatchSnapshot()
})

test("should support all jsonable values", () => {
  const code = `
    const a = __num__
    const b = __str__
    const c = __bool__
    const d = __null__
    const e = __undefined__
    const f = __function__
    const e = __array__
    const g = __object__
    const h = __regex__
  `
  const r = transform(code, {
    macros: [
      createLitMacro("__num__", 123),
      createLitMacro("__str__", "Hello World"),
      createLitMacro("__bool__", false),
      createLitMacro("__null__", null),
      createLitMacro("__undefined__", undefined),
      createLitMacro("__function__", () => true, "() => true"),
      createLitMacro("__array__", [1, 2, 3], "number[]"),
      createLitMacro("__object__", { a: 1, b: 2 }, "Record<string, number>"),
      createLitMacro("__regex__", /.*/g),
    ],
    emitDts: true
  })

  expect(r.code).toMatchSnapshot()
  expect(r.dts).toMatchSnapshot()
})

test("should allow ast for more complex value", () => {
  const code = `
    const expr = __expr__
  `
  expect(transform(code, {
    macros: [
      createLitMacro("__expr__", parseExpr('call("hello", "world")'))
    ]
  }).code).toMatchSnapshot()
})

test("should support pass an object", () => {
  const code = `
    const a = __num__
    const b = __str__
    const c = __bool__
    const d = __null__
    const e = __undefined__
    const f = __function__
    const e = __array__
    const g = __object__
    const h = __regex__
    const i = __expr__
  `
  const r = transform(code, {
    macros: [
      createLitMacro({
        __num__: 123,
        __str__: "Hello World",
        __bool__: false,
        __null__: null,
        __undefined__: undefined,
        __function__: () => true,
        __array__: [1, 2, 3],
        __object__: { a: 1, b: 2 },
        __regex__: /.*/g,
        __expr__: parseExpr('call("hello", "world")')
      }, {
        __function__: "() => true",
        __array__: "number[]",
        __object__: "Record<string, number>",
        __expr__: "string"
      })
    ],
    emitDts: true
  })
  expect(r.code).toMatchSnapshot()
  expect(r.dts).toMatchSnapshot()
})
