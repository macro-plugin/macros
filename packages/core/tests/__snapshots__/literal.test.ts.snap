// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create literal macro 1`] = `
"if (true) {
    console.log("development");
}
"
`;

exports[`should allow ast for more complex value 1`] = `
"const expr = call("hello", "world");
"
`;

exports[`should pass overwrited variable 1`] = `
"const before = true;
function testDev(__DEV__) {
    if (__DEV__) {
        console.log("development");
    }
}
const after = true;
"
`;

exports[`should support all jsonable values 1`] = `
"const a = 123;
const b = "Hello World";
const c = false;
const d = null;
const e = undefined;
const f = ()=>true;
const e = [
    1,
    2,
    3
];
const g = {
    "a": 1,
    "b": 2
};
const h = /.*/g;
"
`;

exports[`should support pass an object 1`] = `
"const a = 123;
const b = "Hello World";
const c = false;
const d = null;
const e = undefined;
const f = ()=>true;
const e = [
    1,
    2,
    3
];
const g = {
    "a": 1,
    "b": 2
};
const h = /.*/g;
const i = call("hello", "world");
"
`;
